### DISCUSSION TITLE
FFI is pain in the butt

### DISCUSSION THREAD

**Discussion by AlexanderTheGr8-gr** [libvips/php-vips#143] [General]
Hello.

We are a small web development company and we decided to use LibVIPS and PHP Vips upon reading about it's famed performance.
We run on CloudLinux 7.9 (Boris Yegorov).
We have a WHM with EasyAppache4 installed, available software to install PECL and PEAR packages with one click,
MultiPHP ini editor, etc
We have installed LibVIPS, PHPVIPS(2.x), and were struggling, yes struggling, to make FFI play...

FFI is indeed a record in our phpinfo, but it only reports the creators, nothing more.

> Module | Authors
> FFI | Dmitry Stogov

</body>
</html>

As php [manual page](https://www.php.net/manual/en/ffi.installation.php) informs, 

> To enable the FFI extension, PHP has to be configured with --with-ffi

Now, all the versions of PHP we can get is from EasyAppache4,
and they don't seem to be configured "--with-ffi"...

Also searching on "building PHP", I found it seems a rather abstract procedure, I didn't find very specific instructions,
with a quick search it seems that this process for a newcomer on building phps from source has a long learning time.

I can see of course how it makes the work of the one implementing the library easier and quicker,
but on the other hand, for anyone not having a PHP preconfigured with "--with-ffi", this is a dead-end.

Another concern is the one stated at a [site](https://thephp.website/en/issue/php-ffi/), that with a c extension the performance is way better than using ffi...

> FFI or C Extensions, which should you use?
> FFI is often given as a tool for prototyping: you make your first steps with it and later on migrate to a native extension code.
> 
> I think if your code doesn't care much about performance (unlikely, but can happen) it is okay to use FFI just to extend PHP's capability

Therefore I wander if it truly is a better option to use FFI, than a c extension.

If you still believe it's a good option (it doesn't concern me, or our company, it's your choice and understanding),
we would be grateful if you could suggest another way around.
Currently we changed our plan to switch back to PHP VIPS 1.x with the ext, hoping nothing will go wrong, we lost some tome already trying to make the project with libvips work..

Any kind of suggestion,
from page with instructions on building php 8 an easy way (without needing to rebuild appache),
or at least instructions in a more specific way on building php,
to instructions on how or where to actually find a php version prebuilt "--with-ffi" for our O.S.
(in order for us to throw it at the folders inside WHM),
would be greatly appreciated.

Thank you for your work,
it would be good for our company to actually reach the point to verify those performance tests :)
Have a good day.

**Comment by jcupitt** [comment]
Hi @alexNecroJack,

Thanks for trying php-vips 2.x! I'm sorry it didn't work out. 

I think the whole of pecl and pear is on life support. There was talk of removing all of it for php 8, but somehow it has lingered on. It's poorly maintained, and things like Windows support just aren't there. The php-vips-ext extension was annoying to maintain, and caused users a lot of trouble. 

I hoped php-ffi would fix all these problems: it should be installed for users (so we wouldn't have to do that), should work everywhere (even on windows!), and it would mean we could write the whole of php-vips in pure PHP. I would expect it to be slightly slower, but only very slightly.

I think your PHP is probably built with ffi (it's the default). Maybe it just needs enabling?

Do you use docker? Can you share a dockerfile that would help me understand your platform? I have some samples here:

https://github.com/jcupitt/docker-builds

**Reply by jcupitt** [reply]
Ah, it's cpanel. I've tried signing up for a trial, they've not mailed me back yet.

I think you have ffi, you maybe just need to enable it. Add this to your `php.ini`:

```
ffi.enable true
```

See:

https://www.php.net/manual/en/ffi.configuration.php

**Reply by AlexanderTheGr8-gr** [reply]
Hello @jcupitt .

Thank you for all the info around those maters I have no knowledge of.

I will have to insist though, because as I have mentioned before, it's not only necessary to enable it,
it is stated in [PHP manual](https://www.php.net/manual/en/ffi.installation.php) that the PHP version anyone uses must be (pre)configured to be "--with-ffi"

Here are some screenshots:
1 - [PHP info - partial](https://enasmaragos.gr/download/Untitled.png)
2 - [PHP ini - partial](https://enasmaragos.gr/download/Untitled2.png)
3 - [The error that gets produced](https://enasmaragos.gr/download/Untitled3.png)

1 - As you can see, searching with-ffi returns no results, all the configuration is there. Also ffi returns only the row displayed in first screenshot about it's author.
2 - I tried both the ini lines, just in case (one by one) to make sure it's not syntactic mistake
3 - You see it's stated in an obvious way, there is no class.

The code I used is:
```PHP
<?php
  echo "<pre>".file_get_contents('/opt/cpanel/ea-php80/root/etc/php.ini')."</pre>";
  //die();
  
  ini_set('display_errors', 1);
  ini_set('display_startup_errors', 1);
  error_reporting(E_ALL);
  
?><?php
  echo 'hi';
  // function_exists();
  $ffi = \FFI::cdef(
      "int printf(const char *format, ...);", // this is a regular C declaration
      "libc.so.6");
  // call C's printf()
  $ffi->printf("Hello %s!\n", "world");
  echo ' hello';
```

If you do find any info regarding what is missing or what to do in WHM / CPanel, please let me know.
I'll also add another note, that according to [manual](https://www.php.net/manual/en/ffi.requirements.php)
> libffi library has to be installed
We did install it, but it's not on PATH, I don't know if that could play a part in this.
```
yum install libffi libffi-devel
//And although it reports libffi-3.0.13-19.el7.x86_64 got installed, the line below finds it not
which libffi-3.0.13-19.el7.x86_64
```

Either way, thank you for your time!

**Reply by AlexanderTheGr8-gr** [reply]
I have search all around for what can be done. I reach to the point that I can do nothing.

While searching with yum, I found one, single package, for ffi --> alt-php80-ffi.x86_64
Of course this is not bundled with PHP so it wouldn't work even if I installed it.

I learned that EasyApache4 has it's own packages for consumption.
It has a way to install "Vendor-provided versions of PHP" and even build,
but that's way out of my water pond.

Is there any place where a package exists for PHP build with ffi,
for an O.S. as described above (CloudLinux 7.9 (Boris Yegorov)),
that can be installed and be detected for use from EasyApache4?

Or is there a way to build PHP without building Apache, and use it in EasyApache4?

I'm asking for directions, not full written solutions. Any search phrase that you have seen bears fruits would be valuable,
Most searches I do on the matter return fruitless...
thank you.

**Reply by jcupitt** [reply]
I think I'd need to experiment with an easyapache install (I've never used it), but they've not answered my application for a test license :( I'll try asking them again.

**Reply by jcupitt** [reply]
fwiw, if I do `php -i` (the command-line equivalent of `phpinfo()`) I see:

```
$ php -i
....
FFI

FFI support => enabled

Directive => Local Value => Master Value
ffi.enable => preload => preload
ffi.preload => no value => no value
```

That's for FFI installed and enabled. If I disable FFI in my `php.ini` I see:

```
FFI

FFI support => enabled

Directive => Local Value => Master Value
ffi.enable => Off => Off
ffi.preload => no value => no value
```

That's php 8.1. 

As a workaround (perhaps until I understand what easyapache are doing with FFI), I suggest you stick with php-vips 1.x. I wasn't planning to add any more features, but it'll still work, and you can still use it with newer libvipses.

**Reply by kleisauke** [reply]
CloudLinux 7.9 seems to be based on RHEL 7 so I assume you could use Remi's RPM repository, a third-party repository that includes an up-to-date version of libvips and PHP. See https://github.com/libvips/libvips/issues/1184#issuecomment-447973135.

```bash
# Install and enable the EPEL RPM package on CloudLinux 7
yum install https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm

# Install the Remi repository configuration package
yum install http://rpms.remirepo.net/enterprise/remi-release-7.rpm

# Install the yum-utils package (for the yum-config-manager command)
yum install yum-utils

# Command to enable the repositories
yum-config-manager --enable remi
yum-config-manager --enable remi-php81

# Install libvips (+ development files and command-line tools)
yum install vips vips-devel vips-tools

# Install PHP 8.1 (+ command-line interface and FFI)
yum install php php-cli php-ffi
```

https://rpms.remirepo.net/enterprise/7/php81/x86_64/repoview/php-ffi.html
